builtins: implement certain Geometry extraction functions

otan opened the pull request:
Gets us over the Chapter 10 line.

Release note (sql change): Implement the following builtins:
* ST_X (resolves #49069)
* ST_Y (resolves #49070)
* ST_NDims (resolves #48992)
* ST_NumPoints (resolves #49001)
* ST_NPoints (resolves #48995)
* ST_StartPoint (resolves #49047)
* ST_EndPoint (resolves #48924)
* ST_NumInteriorRings (resolves #48999)
* ST_NumInteriorRing (resolves #48998)
* ST_InteriorRingN (resolves #48949)
* ST_ExteriorRing (resolves #48930)
* ST_NumGeometries (resolves #48997)
* ST_GeometryN (resolves #48945)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49341)
<!-- Reviewable:end -->

otan committed:
builtins: implement certain Geometry extraction functions

Gets us over the Chapter 10 line.

Release note (sql change): Implement the following builtins:
* ST_X (resolves #49069)
* ST_Y (resolves #49070)
* ST_NDims (resolves #48992)
* ST_NumPoints (resolves #49001)
* ST_NPoints (resolves #48995)
* ST_StartPoint (resolves #49047)
* ST_EndPoint (resolves #48924)
* ST_NumInteriorRings (resolves #48999)
* ST_NumInteriorRing (resolves #48998)
* ST_InteriorRingN (resolves #48949)
* ST_ExteriorRing (resolves #48930)
* ST_NumGeometries (resolves #48997)
* ST_GeometryN (resolves #48945)

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-:-M7xpDwZ-x3tTe8v6KNd:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 728 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xn9NODh1wSO24cgIY:-M7xn9NODh1wSO24cgIZ:b-ijy01r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L728)):*
> ```Go
> 					coord := t.Coord(0)
> 					retG, err := geo.NewGeometryFromGeom(
> 						geom.NewLineStringFlat(geom.XY, []float64{coord.X(), coord.Y()}).SetSRID(t.SRID()),
> ```

why is this a linestring and not a point?
https://postgis.net/docs/ST_StartPoint.html also says point
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xnPy_64aYOlsFktWK:-M7xnPy_64aYOlsFktWL:b-l8l7r9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L759)):*
> ```Go
> 					coord := t.Coord(t.NumCoords() - 1)
> 					retG, err := geo.NewGeometryFromGeom(
> 						geom.NewLineStringFlat(geom.XY, []float64{coord.X(), coord.Y()}).SetSRID(t.SRID()),
> ```

same question
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xo4eP3n72uETUtMbD:-M7xo4eP3n72uETUtMbE:b4crokz) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L874)):*
> ```Go
> 			},
> 			Info: infoBuilder{
> 				info: `Returns the n-th interior ring of a Polygon as a LineString. Returns NULL if the shape is not a Polygon, or the ring does not exist.`,
> ```

Can you add that n is 1-indexed to the `info`.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 886 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xogxw5-6dC9-HeTnZ:-M7xogxw5-6dC9-HeTn_:blayil4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L886)):*
> ```Go
> 			Fn: func(_ *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
> 				g := *args[0].(*tree.DGeometry)
> 				n := int(*args[1].(*tree.DInt))
> ```

this remaining code may read simpler by doing
`n := int(*args[1].(*tree.DInt)) - 1`

and can you add that n is 1-based to the info string.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49341)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/sql/sem/builtins/geo_builtins.go, line 728 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xn9NODh1wSO24cgIY:-M7zG_-w0Mckq48zNOdA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L728)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is this a linestring and not a point?
https://postgis.net/docs/ST_StartPoint.html also says point
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 759 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xnPy_64aYOlsFktWK:-M7zG_V5C7fPd85soiHh:b-8gb715) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L759)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

same question
</blockquote></details>

Done. Mistake - thanks for catching.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 874 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xo4eP3n72uETUtMbD:-M7zGqJl8ztchV5983UQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L874)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you add that n is 1-indexed to the `info`.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 886 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49341#-M7xogxw5-6dC9-HeTnZ:-M7zGpvTDv5IVHq7q0Az:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ef3727e891b9ffd89be594d1a1b27bac9cf1aa0/pkg/sql/sem/builtins/geo_builtins.go#L886)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this remaining code may read simpler by doing
`n := int(*args[1].(*tree.DInt)) - 1`

and can you add that n is 1-based to the info string.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
